skbio/diversity/beta/_fast_unifrac.py

```diff
@@ -0,0 +1,444 @@
+     - unnormalized_unifrac
+     - G
+     - unnormalized_G
+    """
+
+    _or, _and = np.logical_or(i, j), np.logical_and(i, j)
+    return 1 - ((branch_lengths * _and).sum() / (branch_lengths * _or).sum())
+
+
+def w_unifrac(branch_lengths, tip_indices, i, j):
+    """Calculates weighted unifrac(i,j) from branch lengths and cols i,j of m.
+    """
+    i_sum = (np.take(i, tip_indices)).sum()
+    j_sum = (np.take(j, tip_indices)).sum()
+
+    if i_sum:
+        i_ = i / i_sum
+    else:
+        i_ = 0.0
+
+    if j_sum:
+        j_ = j / j_sum
+    else:
+        j_ = 0.0
+
+    return (branch_lengths * abs(i_ - j_)).sum()
+
+
+def _skbio_counts_to_envs(otu_ids, *counts):
+    """This is a place holder method to help get the API hooked up
+
+    This method should be unnecessary once the implementation is optimized to
+    the interface.
+    """
+    envs = {}
+    n_counts = len(counts)
+
+    for packed in zip(otu_ids, *counts):
+        # NOTE: this is ducttape to fit the API
+        otu_id = packed[0]
+
+        counts = {}
+        for env in range(1, n_counts + 1):
+            if packed[env]:
+                counts[env] = packed[env]
+
+        if counts:
+            envs[otu_id] = counts
+
+    return envs
+
+
+def unweighted_unifrac(u_counts, v_counts, otu_ids, tree,
+                       suppress_validation=False):
+    """fit to unifrac API"""
+    if not suppress_validation:
+        _validate(u_counts, v_counts, otu_ids, tree)
+
+    u_sum = sum(u_counts)
+    v_sum = sum(v_counts)
+
+    if not u_sum or not v_sum:
+        if u_sum + v_sum:
+            # u or v counts are all zeros
+            return 1.0
+        else:
+            # u and v are zero
+            return 0.0
+
+    envs = _skbio_counts_to_envs(otu_ids, u_counts, v_counts)
+
+    return fast_unifrac(tree, envs)
+
+
+def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+                     suppress_validation=False):
+    u_sum = sum(u_counts)
+    v_sum = sum(v_counts)
+
+    if not u_sum or not v_sum:
+        if u_sum + v_sum:
+            # u or v counts are all zeros
+            if normalized:
+                return 1.0
+            # NOTE: we cannot handle the unnormalized case here yet as it
+            # requires operations on the tree vector
+        else:
+            # u and v are zero
+            return 0.0
+    if not suppress_validation:
+        _validate(u_counts, v_counts, otu_ids, tree)
+    envs = _skbio_counts_to_envs(otu_ids, u_counts, v_counts)
+
+    return fast_unifrac(tree, envs, weighted=True, normalized=normalized)
+
+
+def fast_unifrac(t, envs, weighted=False, metric=unifrac, normalized=False):
+    # weighted_unifrac_f=_weighted_unifrac,make_subtree=True):
+    """
+    Run fast unifrac.
+
+    Parameters
+    ----------
+    t: skbio.TreeNode
+        phylogenetic tree relating the sequences.
+    envs: dict
+        dict of {sequence:{env:count}} showing environmental abundance.
```

Comment:
I'm pretty sure that `sequence` corresponds to the sequence id (i.e. the OTU id).